[TSVB] Fix color rules are not applied for null series

Kuznietsov opened the pull request:
### Fixes #100250.

This PR is :

- changing the way of handling empty data in metrics, top-N and gauge visualizations
- provides a new rule, ***`Empty`***, which is applying color rules for empty series
- If previous visualizations contained rules ***`less then or equal zero`*** or  ***`greater then or equal zero`***, empty color rules are automatically added to a list of rules, based on the last matched rule.
 
---

#### Empty rule in markdown tab:
![Снимок экрана 2021-05-28 в 12 53 35](https://user-images.githubusercontent.com/22456368/119966187-badc8f00-bfb3-11eb-93f1-d22e36b122b8.png)

---

#### Empty rule in Top N tab:
![Снимок экрана 2021-05-28 в 12 55 10](https://user-images.githubusercontent.com/22456368/119966398-f1b2a500-bfb3-11eb-999c-b1f1660669cb.png)

---

#### Empty rule in Gauge tab:
![Снимок экрана 2021-05-28 в 12 55 50](https://user-images.githubusercontent.com/22456368/119966493-08f19280-bfb4-11eb-92c1-a919bde44b97.png)

---

#### Migrations:
##### If some visualization contained ***`less then or equal zero`*** or  ***`greater then or equal zero`*** color rules, an empty color rule will be added automatically, based on the last matched rule.

---

#### Empty rule, added during migration:
![Снимок экрана 2021-05-28 в 13 05 38](https://user-images.githubusercontent.com/22456368/119967799-689c6d80-bfb5-11eb-8719-92ec73ce8377.png)

Kuznietsov committed:
Fixed color rules behaviour on empty metrics data.

Refactored function `getLastValue`, which was providing unexpected result on empty arrays. It was returning string, instead of null/undefined value.
Created two useful functions, which are providing possibility to handle empty result and get default value, which is expected.

Kuznietsov committed:
Tests added.

Tests for getLastValue refactored.
Tests for getLastValueOrDefault and getLastValueOrZero written.

Kuznietsov committed:
Removed console.log

Kuznietsov committed:
Added default value for empty operator.

Added default value for empty operator, which will compare statistics to empty array.
Added conditional render of colorRuleValue field, if operator doesn't require some specified value to be chosen ( as default, in this case ).

Kuznietsov committed:
Added empty data handling.

Added empty value var and way of displaying in widgets.
Added way of handling empty results and prevented comparing null, empty array and numeric rules.

Kuznietsov committed:
Prettier fixes.

Kuznietsov committed:
Added the same logic of displaying data to gauge.

Added displaying of empty data to gauge module.
Fixed label color styles (before, it was ignoring, because of setting colorValue out of default scope while reactcss(...) call).

Kuznietsov committed:
Added empty data handling in Top N chart.

Kuznietsov committed:
Removed getLastValueOrZero.

Removed getLastValueOrZero and replaced by getLastValueOrEmpty.

Kuznietsov committed:
Added isEmptyValue function.

Added isEmptyValue function, which is able to check equality. It provides a possibility to encapsulate the logic of empty value and flexible changing of its' behavior.

Kuznietsov committed:
Fixed and refactor.

Fixed hidden value input, if no operator selected.
Removed useless DEFAULT_VALUE and getLastValueOrDefault.

Kuznietsov committed:
Color rules Tests.

Changed from js to ts last_value_utils. Updated tests for color_rules component.

Kuznietsov committed:
Replaces isEqual rule with eq.

Kuznietsov committed:
Migrations added.

kibanamachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

alexwizp commented:
Jenkins, test please

alexwizp commented on src/plugins/visualizations/server/saved_objects/visualization_migrations.ts:
```diff
@@ -981,6 +982,81 @@ const removeDefaultIndexPatternAndTimeFieldFromTSVBModel: SavedObjectMigrationFn
+      const gaugeEmptyRules = convertToArrayOfEmtpyRules(lastGaugeRule);
+
+      const colorRules = {
+        bar_color_rules: [...barColorRules, ...barEmptyRules],
```
it seems to me that it is correct to change something only if we have found a rule that needs to be replaced

alexwizp commented on src/plugins/visualizations/server/saved_objects/visualization_migrations.ts:
```diff
@@ -981,6 +982,81 @@ const removeDefaultIndexPatternAndTimeFieldFromTSVBModel: SavedObjectMigrationFn
+      const lastBackgroundRule = getRulesWithComparingToZero(backgroundColorRules) ?? null;
+      const lastGaugeRule = getRulesWithComparingToZero(gaugeColorRules) ?? null;
+
+      const barEmptyRules = convertToArrayOfEmtpyRules(lastBarRule);
```
if we want to replace all `not in array rules` it should be done in a separe migration script + code should  be updated to work only with arrays.

alexwizp commented on src/plugins/visualizations/server/saved_objects/visualization_migrations.ts:
```diff
@@ -981,6 +982,81 @@ const removeDefaultIndexPatternAndTimeFieldFromTSVBModel: SavedObjectMigrationFn
+      const getRulesWithComparingToZero = (rules: any[] = []) =>
+        last(
+          rules.filter(
+            (rule) => compareWithEqualMethods.indexOf(rule.operator) !== -1 && rule.value === 0
```
nit: compareWithEqualMethods.includes

alexwizp commented on src/plugins/vis_type_timeseries/common/last_value_utils.ts:
```diff
@@ -0,0 +1,31 @@
+  return extractValue(last(data));
+};
+
+type GetLastValueOrEmptyFn<T, U> = (data: T[][] | void) => T | U;
```
usually we are not writing types for each function. It doesn't have a lot of sense because you can do specify all declaration directly for function.
But if for some reason you need a type for this, declare it like this:  `type ExtractValueFn = typeof getLastValueOrEmpty;`

Kunzetsov replied:
I don't really need types for current functions. I'll remove type definitions. I had to change a file extension to ts because of its import to other ts files. Didn't want to leave it in simple js form)

alexwizp commented on src/plugins/vis_type_timeseries/public/application/components/color_rules.tsx:
```diff
@@ -100,9 +120,14 @@ export class ColorRules extends Component<ColorRulesProps> {
 
   handleOperatorChange = (item: ColorRule) => {
     return (options: Array<EuiComboBoxOptionOption<string>>) => {
+      const selectedOperator: ColorRulesOperator | void = colorRulesOperatorsList.find(
```
nit: `colorRulesOperatorsList` already typed. No need to add extra `ColorRulesOperator | void`

![image](https://user-images.githubusercontent.com/20072247/119118058-82c4d180-ba32-11eb-843e-d10d2711d603.png)

alexwizp commented on src/plugins/vis_type_timeseries/public/application/components/color_rules.tsx:
```diff
@@ -100,9 +120,14 @@ export class ColorRules extends Component<ColorRulesProps> {
+      const selectedOperator: ColorRulesOperator | void = colorRulesOperatorsList.find(
+        (operator) => options[0]?.value === operator.method
+      );
+      const value = (selectedOperator && selectedOperator.value) ?? null;
```
nit: for some places you use `options[0]?.value` for some`(selectedOperator && selectedOperator.value)`
I vote for one style =)

alexwizp commented on src/plugins/vis_type_timeseries/public/application/components/color_rules.tsx:
```diff
@@ -19,8 +19,10 @@ import {
 import { FormattedMessage } from '@kbn/i18n/react';
 import { i18n } from '@kbn/i18n';
 
+import { fatalErrorsServiceMock } from 'src/core/public/mocks';
```
;) something goes wrong here

alexwizp commented on src/plugins/vis_type_timeseries/common/last_value_utils.ts:
```diff
@@ -0,0 +1,31 @@
+
+type GetLastValueOrEmptyFn<T, U> = (data: T[][] | void) => T | U;
+export const getLastValueOrEmpty: GetLastValueOrEmptyFn<unknown, []> = (data) =>
+  getLastValue(data) ?? EMPTY_VALUE;
```
this part is not entirely clear to me. I can see that `getLastValue(data)` is returning a simple value. Why EMPTY_VALUE is array?

Kunzetsov replied:
I needed some value, which would be clearly explaining, that data was empty and need to handle in a different way, than usual.  Sure, I can change it to null.

alexwizp commented on src/plugins/vis_type_timeseries/public/application/components/vis_types/gauge/vis.js:
```diff
@@ -10,9 +10,10 @@ import PropTypes from 'prop-types';
 import { Gauge } from '../../../visualizations/views/gauge';
-import { getLastValue } from '../../../../../common/get_last_value';
+import { getLastValueOrEmpty } from '../../../../../common/last_value_utils';
+const OPERATORS = { gt, gte, lt, lte, eq: isEqual };
```
what do you think about using `isNull/isEmpty` instead of isEqual. I think it that case from `getLastValue` method we can return null instead of empty array

Kunzetsov replied:
I dislike, that it will change the way of handling values. Right now all functions are taking 2 parameters. If we will add isNull/isEmpty, it will pass normally, but that is not good if we know, that some function will ignore our arguments.

But I can do it, no problem, as it will not cause any harm)

alexwizp commented on src/plugins/visualizations/server/saved_objects/visualization_migrations.ts:
```diff
@@ -1021,4 +1097,5 @@ export const visualizationSavedObjectTypeMigrations = {
     hideTSVBLastValueIndicator,
     removeDefaultIndexPatternAndTimeFieldFromTSVBModel
   ),
+  '7.13.1': flow(addEmptyValueRuleForSavedObjectsWithLessAndGreaterThenZeroRules),
```
please add some unit tests for that

Kuznietsov committed:
Fixed types, EMPTY_VALUE, empty method.

Removed type definition for methods in last_value_utils.ts.
Changed EMPTY_VALUE from array to null. Removed default value.
Added logic for handling empty value handling and comparison methods.

Kuznietsov committed:
Fixed comparing null and numeric rules.

Kuznietsov committed:
Changed migrations.

Kuznietsov committed:
Added test for migrations.

Kunzetsov commented:
@alexwizp, I've changed the code based on your suggestions. Please, review it again when it is comfortable. Thanks.

Kuznietsov committed:
Migration fix.

alexwizp commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'master' into #100250

stratoula commented:
Jenkins, test this

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/126845/)
* Commit: 603bec127a48abdb3b8d5a8af431843d15444dec
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [Check i18n](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/126845/execution/node/432/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/126845/","number":"126845","commit":"603bec127a48abdb3b8d5a8af431843d15444dec"}]}
PIPELINE-->

alexwizp commented on src/plugins/vis_type_timeseries/common/last_value_utils.ts:
```diff
@@ -0,0 +1,26 @@
+  return extractValue(last(data));
+};
+
+export const getLastValueOrEmpty = (data: unknown) => getLastValue(data) ?? EMPTY_VALUE;
```
nit: just wondering do we need that method now? Before you worked with an `array` but now we use `null` instead. 
So now `getLastValue` returns `null` or `undefined` for all case.... Please think about it, but I don't insist to fix that

alexwizp commented on src/plugins/visualizations/server/saved_objects/visualization_migrations.ts:
```diff
@@ -981,6 +982,96 @@ const removeDefaultIndexPatternAndTimeFieldFromTSVBModel: SavedObjectMigrationFn
+        id: uuid.v4(),
+      });
+
+      const lastBarRule = getRulesWithComparingToZero(barColorRules) ?? null;
```
nit: extra `?? null `

alexwizp commented on src/plugins/visualizations/server/saved_objects/visualization_migrations.ts:
```diff
@@ -981,6 +982,96 @@ const removeDefaultIndexPatternAndTimeFieldFromTSVBModel: SavedObjectMigrationFn
+
+      const convertRuleToEmpty = ({ id, ...rest }: any = {}) => ({
+        ...rest,
+        ...emptyRule,
```
nit: what do you think about moving `emptyRule -> convertRuleToEmpty` and `compareWithEqualMethods -> getRulesWithComparingToZero`.  
It seems to me that if we already have functions responsible for something, all associated code should be inside

alexwizp commented on src/plugins/visualizations/server/saved_objects/visualization_migrations.ts:
```diff
@@ -981,6 +982,96 @@ const removeDefaultIndexPatternAndTimeFieldFromTSVBModel: SavedObjectMigrationFn
+        gauge_color_rules: gaugeColorRules,
+      };
+
+      if (lastBarRule) {
```
nit: maybe to avoid this type of duplication we can create a separate method for that?  

```
const addEmptyRule = (rules, key) => {
  const rulesWithComparingToZero = getRulesWithComparingToZero(barColorRules);

  if (rulesWithComparingToZero) {
    colorRules[key] = [...colorRules[key], convertRuleToEmpty(rulesWithComparingToZero)];
  }
};
```

Kunzetsov replied:
We are modifying some external array in this case... As for me, such mutations are very dangerous, because it is difficult to recognize, what is going wrong with the external data structure. Exactly that's why I was modifying the array last time on every step.

Kunzetsov replied:
I've taken your advice and it looks much better right now. Thanks)

Kuznietsov committed:
Updated code, based on nits and fixed reasons of pipeline errors.

Kuznietsov committed:
Merge branch '#100250' of github.com:Kunzetsov/kibana into #100250

VladLasitsa commented on src/plugins/vis_type_timeseries/public/application/components/vis_types/metric/vis.js:
```diff
@@ -22,9 +26,14 @@ function getColors(props) {
+      if (rule.operator) {
+        const value = getLastValue(series[0]?.data);
+        // This check is necessary for preventing from comparing null values with numeric rules.
+        const shouldOperate =
+          (isNull(rule.value) && OPERATORS_ALLOW_NULL[rule.operator]) ||
+          (!isNull(rule.value) && !isNull(value));
```
It occurs in the code several times. Maybe we can create a separate 'check' function for this?

Kunzetsov replied:
Thanks for this suggestion) Sounds reasonable. Possible, even add OPERATORS and everything connected to them to the separate file. We will avoid duplicating code in metrics, top_n and gauge/vis.js files.

Kuznietsov committed:
Moved actions, connected to operators to the separate file. Reduced duplication of code.

Kuznietsov committed:
Type names changed.

Kuznietsov committed:
Test for operators_utils added.

VladLasitsa reviewed (approved):
LGTM

alexwizp commented on src/plugins/vis_type_timeseries/common/operators_utils.ts:
```diff
@@ -0,0 +1,47 @@
+
+import { gt, gte, lt, lte, isNull } from 'lodash';
+
+type Operator = 'gte' | 'lte' | 'gt' | 'lt' | 'empty';
```
nit: I think the code below can be simple converted to simple `enum` 
```
type Operator = 'gte' | 'lte' | 'gt' | 'lt' | 'empty';

export const GTE: Operator = 'gte';
export const LTE: Operator = 'lte';
export const GT: Operator = 'gt';
export const LT: Operator = 'lt';
export const EMPTY: Operator = 'empty';

const OPERATORS = {
  [GTE]: gte,
  [LTE]: lte,
  [GT]: gt,
  [LT]: lt,
  [EMPTY]: isNull,
};
```

alexwizp commented on src/plugins/visualizations/server/saved_objects/visualization_migrations.ts:
```diff
@@ -981,6 +982,78 @@ const removeDefaultIndexPatternAndTimeFieldFromTSVBModel: SavedObjectMigrationFn
+      };
+
+      const colorRules = {
+        bar_color_rules: addEmptyRuleToListIfNecessary(barColorRules),
```
nit: `barColorRules` constant looks a little bit extra for me 

I think the code below does not serve any purpose
```
const {
        bar_color_rules: barColorRules = [],
        background_color_rules: backgroundColorRules = [],
        gauge_color_rules: gaugeColorRules = [],
      } = params;
```

what if we simply pass `params['bar_color_rules']` into `addEmptyRuleToListIfNecessary`

alexwizp commented on src/plugins/visualizations/server/saved_objects/visualization_migrations.ts:
```diff
@@ -981,6 +982,78 @@ const removeDefaultIndexPatternAndTimeFieldFromTSVBModel: SavedObjectMigrationFn
+      const convertRuleToEmpty = (rule: any = {}) => ({
+        ...rule,
+        id: uuid.v4(),
+        operator: 'empty',
```
nit: if I'm not wrong you created a constant for that in common folder. Let's use it

Kunzetsov replied:
I was thinking about it but didn't know if it is a good idea to import something to the migrations from other plugins.

Kuznietsov committed:
Fixed based on nits.

Kuznietsov committed:
Added vis_type_timeseries to tsconfig references.

alexwizp commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'master' into #100250

alexwizp reviewed (approved):
LGTM! No more objections from my side

elasticmachine commented:
Pinging @elastic/kibana-app (Team:KibanaApp)

stratoula commented:
@Kunzetsov I will review this tomorrow but can you add a screenshot to the description?

Kunzetsov commented:
> @Kunzetsov I will review this tomorrow but can you add a screenshot to the description?
Sure, I can) But I'll be able to do it only tomorrow morning.

stratoula commented:
Jenkins, test this

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127438/)
* Commit: 0aeef879fc6b1cfcbde536c7b3cda04b91f31a9d
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127438/execution/node/423/log)

### History
* :broken_heart: [Build #126845](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/126845/) failed 603bec127a48abdb3b8d5a8af431843d15444dec

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/126845/","number":"126845","commit":"603bec127a48abdb3b8d5a8af431843d15444dec"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127438/","number":"127438","commit":"0aeef879fc6b1cfcbde536c7b3cda04b91f31a9d"}]}
PIPELINE-->

Kuznietsov committed:
Merge branch 'master' into #100250

# Conflicts:
#	src/plugins/visualizations/server/migrations/visualization_saved_object_migrations.test.ts
#	src/plugins/visualizations/server/migrations/visualization_saved_object_migrations.ts

stratoula commented on src/plugins/visualizations/server/migrations/visualization_saved_object_migrations.ts:
```diff
@@ -1007,6 +1075,7 @@ export const visualizationSavedObjectTypeMigrations = {
     removeDefaultIndexPatternAndTimeFieldFromTSVBModel
   ),
   '7.13.1': flow(
+    addEmptyValueRuleForSavedObjectsWithLessAndGreaterThenZeroRules,
```
We don't want this for 7.13.1
We want this only for 7.14

Moreover the migration logic has a little bit changed in order to take under consideration the by value visualizations (the panels that are created from the dashboard and not saved to the library). So take a look on what we do for 7.13.1 and apply it for 7.14.  Do not forget the `visualize_embeddable_factory`!

Kunzetsov replied:
Ok, thank you.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127830/)
* Commit: 69269fe74dab7727e18720943816f3dccf744e66
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-100404/69269fe74dab7727e18720943816f3dccf744e66)
* This comment will update when the build is complete

### Failed CI Steps
* [Jest Unit Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127830/execution/node/365/log)

### History
* :broken_heart: [Build #127438](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127438/) failed 0aeef879fc6b1cfcbde536c7b3cda04b91f31a9d
* :broken_heart: [Build #126845](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/126845/) failed 603bec127a48abdb3b8d5a8af431843d15444dec

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/126845/","number":"126845","commit":"603bec127a48abdb3b8d5a8af431843d15444dec"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127438/","number":"127438","commit":"0aeef879fc6b1cfcbde536c7b3cda04b91f31a9d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127830/","number":"127830","commit":"69269fe74dab7727e18720943816f3dccf744e66"}]}
PIPELINE-->

Kuznietsov committed:
Changed version and added migrations.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127836/)
* Commit: 3c12c45742837a13c259c7c75c63f2a2505e661f
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [API Integration Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127836/execution/node/409/log)

### History
* :broken_heart: [Build #127830](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127830/) failed 69269fe74dab7727e18720943816f3dccf744e66
* :broken_heart: [Build #127438](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127438/) failed 0aeef879fc6b1cfcbde536c7b3cda04b91f31a9d
* :broken_heart: [Build #126845](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/126845/) failed 603bec127a48abdb3b8d5a8af431843d15444dec

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/126845/","number":"126845","commit":"603bec127a48abdb3b8d5a8af431843d15444dec"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127438/","number":"127438","commit":"0aeef879fc6b1cfcbde536c7b3cda04b91f31a9d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127830/","number":"127830","commit":"69269fe74dab7727e18720943816f3dccf744e66"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127836/","number":"127836","commit":"3c12c45742837a13c259c7c75c63f2a2505e661f"}]}
PIPELINE-->

Kuznietsov committed:
Small fix in migrations.

Kunzetsov commented:
> @Kunzetsov I will review this tomorrow but can you add a screenshot to the description?

Added screenshots and comments.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127845/)
* Commit: b0058e1168696ab7bffd8fac0b5a722935d94dff
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-100404/b0058e1168696ab7bffd8fac0b5a722935d94dff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/fcaded750fd1c04c713a87c12af612712773f1be) | [after](https://github.com/elastic/kibana/commit/b0058e1168696ab7bffd8fac0b5a722935d94dff) | diff |
| --- | --- | --- | --- |
| `visTypeTimeseries` | 504 | 505 | +1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/fcaded750fd1c04c713a87c12af612712773f1be) | [after](https://github.com/elastic/kibana/commit/b0058e1168696ab7bffd8fac0b5a722935d94dff) | diff |
| --- | --- | --- | --- |
| `visTypeTimeseries` | 1.7MB | 1.7MB | +3.9KB |

### History
* :broken_heart: [Build #127836](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127836/) failed 3c12c45742837a13c259c7c75c63f2a2505e661f
* :broken_heart: [Build #127830](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127830/) failed 69269fe74dab7727e18720943816f3dccf744e66
* :broken_heart: [Build #127438](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127438/) failed 0aeef879fc6b1cfcbde536c7b3cda04b91f31a9d
* :broken_heart: [Build #126845](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/126845/) failed 603bec127a48abdb3b8d5a8af431843d15444dec

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/126845/","number":"126845","commit":"603bec127a48abdb3b8d5a8af431843d15444dec"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127438/","number":"127438","commit":"0aeef879fc6b1cfcbde536c7b3cda04b91f31a9d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127830/","number":"127830","commit":"69269fe74dab7727e18720943816f3dccf744e66"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127836/","number":"127836","commit":"3c12c45742837a13c259c7c75c63f2a2505e661f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127845/","number":"127845","commit":"b0058e1168696ab7bffd8fac0b5a722935d94dff"}]}
PIPELINE-->

stratoula commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'master' into #100250

stratoula commented on src/plugins/vis_type_timeseries/public/application/components/color_rules.tsx:
```diff
@@ -54,33 +62,45 @@ const defaultPrimaryName = i18n.translate('visTypeTimeseries.colorRules.defaultP
+      defaultMessage: 'empty',
+    }),
+    method: Operator.Empty,
+    isValueConstant: true,
```
I would name this flag a little bit different. It has confused me a bit tbh.
You can name it something like `hideValueSelector` or `needValue: false` or something like that.

Kunzetsov replied:
Okey, I'll change it.

stratoula commented on src/plugins/visualizations/server/embeddable/visualize_embeddable_factory.ts:
```diff
@@ -36,6 +37,15 @@ const byValueRemoveDefaultIndexPatternAndTimeFieldFromTSVBModel = (state: Serial
   };
 };
 
+const byValueAddEmptyValueRuleForSavedObjectsWithLessAndGreaterThenZeroRule = (
```
This is a nit but I would make the function names smaller:
byValueAddEmptyValueColorRule, commonAddEmptyValueColorRule etc

stratoula commented on src/plugins/visualizations/server/migrations/visualization_common_migrations.ts:
```diff
@@ -42,3 +45,51 @@ export const commonRemoveDefaultIndexPatternAndTimeFieldFromTSVBModel = (visStat
+    const getRuleWithComparingToZero = (rules: any[] = []) => {
+      const compareWithEqualMethods = ['gte', 'lte'];
+      return last(
+        rules.filter((rule) => compareWithEqualMethods.includes(rule.operator) && rule.value === 0)
```
Can you help me with this? Why do we backport for rules gte/gt zero ?
From my understanding the rules that do not work correctly are the rules that are set below zero, right? Except I am missing something here 🤔

Kunzetsov replied:
The fourth step of bug's reproduction was "Apply a color rule (change the background color for metric less than or equal to 0". So, customer expects to see rules, containing 'equal to zero' applied to empty value. As we've decided to treat empty value and zero value differently, we need this backport fo gte zero in addition to lte zero.

Kunzetsov replied:
I was implementing this suggestion: https://github.com/elastic/kibana/issues/100250#issuecomment-843078163

stratoula replied:
Why is it a bug? If the user has series with value greater than 0 and rule about greater than zero it will work. 
cc @alexwizp

Kunzetsov replied:
But where do you see gt?

Kunzetsov replied:
It works only for gte/lte (>= 0, <=0).

Kuznietsov committed:
Fixes based on review.

elasticmachine commented:
Pinging @elastic/apm-ui (Team:apm)

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

elasticmachine commented:
Pinging @elastic/uptime (Team:uptime)

Kuznietsov committed:
Revert "Fixes based on review."

This reverts commit 35af7b2b6a0c7a44a5d57c991837f9f53a517870.

elasticmachine commented:
Pinging @elastic/uptime (Team:uptime)

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

elasticmachine commented:
Pinging @elastic/apm-ui (Team:apm)

Kuznietsov committed:
Fixes based on review.

stratoula reviewed (approved):
I tested it locally, it works fine.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128128/)
* Commit: 1be7a13d6ac1816b0ff559ff304ae08b6793651f
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-100404/1be7a13d6ac1816b0ff559ff304ae08b6793651f)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/418a3d312ce48fa818c1564508e00c6b328bfdbf) | [after](https://github.com/elastic/kibana/commit/1be7a13d6ac1816b0ff559ff304ae08b6793651f) | diff |
| --- | --- | --- | --- |
| `apm` | 1601 | 1600 | -1 |
| `apmOss` | 5 | 6 | +1 |
| `charts` | 68 | 67 | -1 |
| `fleet` | 457 | 456 | -1 |
| `lens` | 650 | 559 | -91 |
| `lists` | 219 | 217 | -2 |
| `ml` | 1756 | 1755 | -1 |
| `securitySolution` | 2225 | 2221 | -4 |
| `visTypeTimeseries` | 504 | 505 | +1 |
| total |  |  | -99 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/418a3d312ce48fa818c1564508e00c6b328bfdbf) | [after](https://github.com/elastic/kibana/commit/1be7a13d6ac1816b0ff559ff304ae08b6793651f) | diff |
| --- | --- | --- | --- |
| `apmOss` | 4 | 38 | +34 |
| `charts` | 155 | 137 | -18 |
| `core` | 1061 | 1073 | +12 |
| `data` | 3223 | 3264 | +41 |
| `embeddable` | 379 | 381 | +2 |
| `expressions` | 1444 | 1456 | +12 |
| `kibanaReact` | 216 | 218 | +2 |
| `kibanaUtils` | 373 | 385 | +12 |
| `lists` | 143 | 141 | -2 |
| `presentationUtil` | 121 | 123 | +2 |
| `ruleRegistry` | 43 | 41 | -2 |
| total |  |  | +95 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/418a3d312ce48fa818c1564508e00c6b328bfdbf) | [after](https://github.com/elastic/kibana/commit/1be7a13d6ac1816b0ff559ff304ae08b6793651f) | diff |
| --- | --- | --- | --- |
| `apm` | 4.3MB | 4.3MB | -125.0B |
| `canvas` | 1.3MB | 1.3MB | -22.0B |
| `charts` | 56.5KB | 54.5KB | -2.0KB |
| `fleet` | 743.0KB | 737.5KB | -5.6KB |
| `lens` | 1.3MB | 1.1MB | -219.8KB |
| `lists` | 270.7KB | 269.5KB | -1.1KB |
| `ml` | 5.9MB | 5.9MB | -2.4KB |
| `observability` | 461.1KB | 461.2KB | +102.0B |
| `securitySolution` | 6.9MB | 6.9MB | -10.8KB |
| `visTypeTimeseries` | 1.7MB | 1.7MB | +3.3KB |
| `visTypeXy` | 113.7KB | 113.7KB | -11.0B |
| total |  |  | -238.4KB |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/418a3d312ce48fa818c1564508e00c6b328bfdbf) | [after](https://github.com/elastic/kibana/commit/1be7a13d6ac1816b0ff559ff304ae08b6793651f) | diff |
| --- | --- | --- | --- |
| `lists` | 38 | 37 | -1 |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/418a3d312ce48fa818c1564508e00c6b328bfdbf) | [after](https://github.com/elastic/kibana/commit/1be7a13d6ac1816b0ff559ff304ae08b6793651f) | diff |
| --- | --- | --- | --- |
| `apm` | 31.9KB | 32.1KB | +233.0B |
| `apmOss` | 2.4KB | 2.6KB | +186.0B |
| `charts` | 87.4KB | 86.6KB | -827.0B |
| `lens` | 34.4KB | 34.8KB | +413.0B |
| `maps` | 63.6KB | 63.6KB | -11.0B |
| `ml` | 67.1KB | 66.6KB | -457.0B |
| `securitySolution` | 61.2KB | 60.8KB | -390.0B |
| `visDefaultEditor` | 33.9KB | 33.9KB | -33.0B |
| total |  |  | -886.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/418a3d312ce48fa818c1564508e00c6b328bfdbf) | [after](https://github.com/elastic/kibana/commit/1be7a13d6ac1816b0ff559ff304ae08b6793651f) | diff |
| --- | --- | --- | --- |
| `apmOss` | 4 | 38 | +34 |
| `charts` | 186 | 165 | -21 |
| `lists` | 151 | 149 | -2 |
| `ruleRegistry` | 43 | 41 | -2 |
| total |  |  | +9 |

#### async chunk count
| id | [before](https://github.com/elastic/kibana/commit/418a3d312ce48fa818c1564508e00c6b328bfdbf) | [after](https://github.com/elastic/kibana/commit/1be7a13d6ac1816b0ff559ff304ae08b6793651f) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 20 | 19 | -1 |

#### References to deprecated APIs
| id | [before](https://github.com/elastic/kibana/commit/418a3d312ce48fa818c1564508e00c6b328bfdbf) | [after](https://github.com/elastic/kibana/commit/1be7a13d6ac1816b0ff559ff304ae08b6793651f) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 346 | 342 | -4 |


</details>

### History
* :green_heart: [Build #127845](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127845/) succeeded b0058e1168696ab7bffd8fac0b5a722935d94dff
* :broken_heart: [Build #127836](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127836/) failed 3c12c45742837a13c259c7c75c63f2a2505e661f
* :broken_heart: [Build #127830](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127830/) failed 69269fe74dab7727e18720943816f3dccf744e66
* :broken_heart: [Build #127438](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127438/) failed 0aeef879fc6b1cfcbde536c7b3cda04b91f31a9d
* :broken_heart: [Build #126845](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/126845/) failed 603bec127a48abdb3b8d5a8af431843d15444dec

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @Kunzetsov

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/126845/","number":"126845","commit":"603bec127a48abdb3b8d5a8af431843d15444dec"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127438/","number":"127438","commit":"0aeef879fc6b1cfcbde536c7b3cda04b91f31a9d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127830/","number":"127830","commit":"69269fe74dab7727e18720943816f3dccf744e66"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127836/","number":"127836","commit":"3c12c45742837a13c259c7c75c63f2a2505e661f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127845/","number":"127845","commit":"b0058e1168696ab7bffd8fac0b5a722935d94dff"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128128/","number":"128128","commit":"1be7a13d6ac1816b0ff559ff304ae08b6793651f"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128143/)
* Commit: d8e43846b24e94333c103706d61f181c83c12f51
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-100404/d8e43846b24e94333c103706d61f181c83c12f51)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/418a3d312ce48fa818c1564508e00c6b328bfdbf) | [after](https://github.com/elastic/kibana/commit/d8e43846b24e94333c103706d61f181c83c12f51) | diff |
| --- | --- | --- | --- |
| `visTypeTimeseries` | 504 | 505 | +1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/418a3d312ce48fa818c1564508e00c6b328bfdbf) | [after](https://github.com/elastic/kibana/commit/d8e43846b24e94333c103706d61f181c83c12f51) | diff |
| --- | --- | --- | --- |
| `visTypeTimeseries` | 1.7MB | 1.7MB | +3.9KB |

### History
* :green_heart: [Build #128128](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128128/) succeeded 1be7a13d6ac1816b0ff559ff304ae08b6793651f
* :green_heart: [Build #127845](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127845/) succeeded b0058e1168696ab7bffd8fac0b5a722935d94dff
* :broken_heart: [Build #127836](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127836/) failed 3c12c45742837a13c259c7c75c63f2a2505e661f
* :broken_heart: [Build #127830](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127830/) failed 69269fe74dab7727e18720943816f3dccf744e66
* :broken_heart: [Build #127438](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127438/) failed 0aeef879fc6b1cfcbde536c7b3cda04b91f31a9d

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @Kunzetsov

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127438/","number":"127438","commit":"0aeef879fc6b1cfcbde536c7b3cda04b91f31a9d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127830/","number":"127830","commit":"69269fe74dab7727e18720943816f3dccf744e66"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127836/","number":"127836","commit":"3c12c45742837a13c259c7c75c63f2a2505e661f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/127845/","number":"127845","commit":"b0058e1168696ab7bffd8fac0b5a722935d94dff"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128128/","number":"128128","commit":"1be7a13d6ac1816b0ff559ff304ae08b6793651f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128143/","number":"128143","commit":"d8e43846b24e94333c103706d61f181c83c12f51"}]}
PIPELINE-->

timroes commented:
@elasticmachine run elasticsearch-ci/docs

kibanamachine commented:
## 💔 Backport failed

| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  7.x  | Commit could not be cherrypicked due to conflicts |

To backport manually run:
`node scripts/backport --pr 100404`
